sdk/program/src/sysvar/slot_hashes.rs

```diff
@@ -97,30 +100,50 @@ impl SlotHashesSysvar {
     /// Get the position of an entry in the sysvar by its key.
     /// Returns `None` if the key is not found.
     pub fn position(slot: &Slot) -> Result<Option<usize>, ProgramError> {
-        get_pod_slot_hashes().map(|pod_hashes| {
+        Self::get_pod_slot_hashes().map(|pod_hashes| {
             pod_hashes
                 .binary_search_by(|PodSlotHash { slot: this, .. }| slot.cmp(this))
                 .ok()
         })
     }
-}
-
-fn get_pod_slot_hashes() -> Result<Vec<PodSlotHash>, ProgramError> {
-    let mut pod_hashes = vec![PodSlotHash::default(); MAX_ENTRIES];
-    {
-        let data = bytemuck::try_cast_slice_mut::<PodSlotHash, u8>(&mut pod_hashes)
-            .map_err(|_| ProgramError::InvalidAccountData)?;
 
-        // Ensure the created buffer is aligned to 8.
-        if data.as_ptr().align_offset(8) != 0 {
-            return Err(ProgramError::InvalidAccountData);
+    /// Return the slot hashes sysvar as a vector of `PodSlotHash`.
+    pub fn get_pod_slot_hashes() -> Result<Vec<PodSlotHash>, ProgramError> {
+        // First fetch the length of the slot hashes vector.
+        let slot_hash_count = {
+            let mut data = vec![0u8; VEC_LEN_U64_SIZE];
+            get_sysvar(
+                &mut data,
+                &SlotHashes::id(),
+                /* offset */ 0,
+                VEC_LEN_U64_SIZE as u64,
+            )?;
+            usize::from_le_bytes(data.try_into().unwrap())
+        };
```

Comment:
This implementation always does two separate syscalls, but there's only ever really a benefit for test validators that were just started or program-test situations, at the cost of causing worse performance for all real-world cases, where there's always 512 entries.

How about always fetching the full thing (size + max_entries * entry_size) but truncating it if the reported size is smaller?

Reply:
I don't believe we can, unfortunately.

https://github.com/anza-xyz/agave/blob/efdbdc00cea2b89338b2e8e1f9ebb6ca77c8b5fa/programs/bpf_loader/src/syscalls/sysvar.rs#L223-L228

As you can see, the syscall is designed to return an error code (`1`) if you try to over-fetch. This prevents the annoying panic from `copy_from_slice` ("source slice length does not match destination slice length").

So without knowing the length of the underlying data in the sysvar cache ahead of time, that's what we'll run into.

Perhaps the better approach to the testing footgun is to add a patch to program-test and/or banks client, where `set_sysvar` preserves the account length?

Reply:
If you send the slot hashes account to a program and log its length in program-test or get the account in a test validato right after starting, you'll always get 20488. Sysvar accounts are padded out to their max length every time: https://github.com/anza-xyz/agave/blob/9a7bf72940f4b3cd7fc94f54e005868ce707d53d/sdk/src/account.rs#L682

This means it's impossible to overfetch.

Reply:
Interesting. I was under the impression the account was overwritten in program-test, but it appears you are correct. I just ran a test.

If you add more than 512 slot hashes to the sysvar, it will extend the account, but I think this is OK.

The only way to totally fudge the account data for the `SlotHashes` sysvar is to literally overwrite the account with `set_account`. At that point, we can't save you.